common/middleware/auth.js

```diff
@@ -0,0 +1,137 @@
+const passport = require('passport')
+const { Strategy } = require('passport-oauth2')
+const refresh = require('passport-oauth2-refresh')
+const { getToken } = require('../data/oauth')
+const config = require('../config')
+const logger = require('../logging/logger')
+
+// Required by Passport - http://www.passportjs.org/docs/configure/#sessions
+passport.serializeUser((user, done) => done(null, user))
+passport.deserializeUser((user, done) => done(null, user))
+
+const tokenVerifier = async (req, enabled = config.apis.oauth.verifyToken) => {
+  const { user, verified } = req
+
+  if (!enabled) {
+    return true
+  }
+
+  if (verified) {
+    return true
+  }
+
+  const result = await getToken(user.token)
+  if (result) {
+    req.verified = true
+  }
+  return result
+}
+
+const checkUserIsAuthenticated = (verifyToken = tokenVerifier) => {
+  return async (req, res, next) => {
+    if (req.isAuthenticated() && (await verifyToken(req))) {
+      return next()
+    }
+
+    req.session.returnUrl = req.originalUrl
+    return res.redirect('/login')
+  }
+}
+
+const userHasExpiredToken = (tokenExpiryTime, nowInSeconds = Date.now()) => tokenExpiryTime < nowInSeconds
```

Comment:
should we include the lower boundary tokenExpiryTime <= nowInSeconds, in case it is just expiring right now at this same second? very very low chance thought.

Reply:
Good spot! - I'll update the assertion to be inclusive :)